### ISSUE TITLE
NuShell install size

### ISSUE THREAD

**Issue by albert-tomanek** [nushell/nushell#13603]
### Related problem

I'd like to install NuShell on an embedded Linux system where disk space is limited. While the `bash` executable on my PC only has 1.3MB, the `nu` executable has 45MB.

### Describe the solution you'd like

I suspect this is because many things are hard-linked in. Is there any way I could compile a more minimal version of `nu`, perhaps with more limited features? The main things I need are the basic commands and the script interpreter.

### Describe alternatives you've considered

Different shells, but the elegance of the `nu` language really appeals to me.

**Comment by fdncred** [comment]
i don't think sh nor bash have > 400 internal commands. that's probably the biggest difference.

you could try changing the optimization setting to optimize for size and recompiling. you could also try modifying the nushell source code to not include the nu-command crates, where most of the dependencies are.

**Comment by albert-tomanek** [comment]
> not include the nu-command crates

Ahh, that sounds like a good starting point. Are there any big (binary-wise) 3rd party libraries that nu links in?

**Comment by fdncred** [comment]
I'm sure there are. You'll just have to play around.

**Comment by KAAtheWiseGit** [comment]
I was thinking about this issue some time ago.  I wonder if it would make sense to have feature flags for Nushell, which disable certain sets of commands.  The default release would have all of them, but it would allow building a more lightweight distribution, without, for example, filesystem/network/plugin/platform commands.

**Comment by fdncred** [comment]
Some time ago we did some crateification work. This is how we broke it down.

### nu-cmd-lang
https://github.com/nushell/nushell/blob/e690e7aac03f80394a444606a45da3bb8ae90dd0/crates/nu-cmd-lang/src/default_context.rs#L17-L64

### nu-cmd-cli
https://github.com/nushell/nushell/blob/e690e7aac03f80394a444606a45da3bb8ae90dd0/crates/nu-cli/src/commands/default_context.rs#L14-L26

### nu-cmd-extra
https://github.com/nushell/nushell/blob/e690e7aac03f80394a444606a45da3bb8ae90dd0/crates/nu-cmd-extra/src/extra/mod.rs#L31-L91


I think that leaves `nu-cmd-plugin` for plugins and `nu-command` for the rest of the commands.

**Comment by sholderbach** [comment]
We explicitly removed the `extra` feature not too far back (#12140)
This also due to the downsides we encountered with people noticing commands not being available, additional friction during development etc.
At the time the number of commands that we actually had in `nu-cmd-extra` was minor compared to things in `nu-command`. 
Then at some point our build pipeline was switched to ship releases with `extra`.
After that I removed that extra feature.

To some extent I prefer us to focus on one high-quality product and don't take on too much configuration that then suffers from reduced testing and development attention. Nevertheless I think we should be frugal with binary size and non-critical bells and whistles that could reasonably land in a plugin. I hoped that we get into a place where there is some force to remove baggage instead of constant growth of new features with potentially large externalities that seemed cool on a whim.

**Comment by KAAtheWiseGit** [comment]
I fully agree.  I don't think that disabling only some of the commands makes sense: I think it should be all or nothing deal.  This way there are no half-states, where some generic scripts work, but others don't.  But I also wish the default version of Nu wouldn't be in tens of megabytes.  On my machine, `uutils` + Fish + cURL total 6.5MB, versus Nu's 18.3MB (all stripped).

**Comment by fdncred** [comment]
You can change the Cargo.toml by updating the release build profile.
```toml
[profile.release]
opt-level = "z"
strip = true
lto = true
codegen-units = 1
panic = 'abort'
```
It took my binary from `37,798,336` to `13,783,888`.
You could also experiement with upx which compresses executables.

**Comment by sholderbach** [comment]
I repeat `panic = 'abort'` is disallowed with our current design including panic hooks (and some loadbearing destructors) (even though panic code is probably a significant contributor to binary size here)

**Comment by KAAtheWiseGit** [comment]
Oh, and I just remembered one thing I wanted to bring up.  Currently Nushell depends on `clap` through uutils crates.  I remember there being a workaround against that, but it seems that it didn't work.

**Comment by fdncred** [comment]
ugh, we kicked clap to the curb long ago because it was a huge dependency and slow to compile. guess we need to keep it now.

**Comment by sholderbach** [comment]
I am guessing most of the `clap` stuff from the uutils gets thrown out during linking, so its more a compile time burden than affecting the install size.
